pkg/stats/historical.go

```diff
@@ -0,0 +1,110 @@
+package stats
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/fastly/cli/pkg/common"
+	"github.com/fastly/cli/pkg/compute/manifest"
+	"github.com/fastly/cli/pkg/config"
+	"github.com/fastly/go-fastly/fastly"
+)
+
+const statusSuccess = "success"
+
+// HistoricalCommand exposes the Historical Stats API.
+type HistoricalCommand struct {
+	common.Base
+	manifest manifest.Data
+
+	Input      fastly.GetStatsInput
+	formatFlag string
+}
+
+// NewHistoricalCommand is the "stats historical" subcommand.
+func NewHistoricalCommand(parent common.Registerer, globals *config.Data) *HistoricalCommand {
+	var c HistoricalCommand
+	c.Globals = globals
+
+	c.CmdClause = parent.Command("historical", "View historical stats for a Fastly service")
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+
+	c.CmdClause.Flag("from", "From time, accepted formats at https://docs.fastly.com/api/stats#Range").StringVar(&c.Input.From)
+	c.CmdClause.Flag("to", "To time").StringVar(&c.Input.To)
+	c.CmdClause.Flag("by", "Aggregation period (minute/hour/day)").EnumVar(&c.Input.By, "minute", "hour", "day")
+	c.CmdClause.Flag("region", "Filter by region ('stats regions' to list)").StringVar(&c.Input.Region)
+
+	c.CmdClause.Flag("format", "Output format (json)").EnumVar(&c.formatFlag, "json")
+
+	return &c
+}
+
+// Exec implements the command interface.
+func (c *HistoricalCommand) Exec(in io.Reader, out io.Writer) error {
+	service, source := c.manifest.ServiceID()
+	if source != manifest.SourceUndefined {
```

Comment:
This would still allow for an empty service ID to be sent in the request. Therefore we should error if none passed and always set the input field.

```suggestion
if source == manifest.SourceUndefined {
    return errors.ErrNoServiceID
}
c.Input.Service = service
```

Reply:
That's intentional, an empty service is a valid request: it will dump all of the services that api token  can read.

Reply:
:+1: Makes sense, however I now feel uncomfortable that the "dump all services" behaviour is essentially an undocumented side-effect and not a first-class discoverable feature - which it should be. 

Also the possibility of `source` being undefined is reduced for C@E services as the service-id is always persisted and read from the `fastly.toml`, which is why we have this pattern as it allows you to omit the `--service-id` flag within your project directory.

So a couple of ideas for the future, but don't need to block this PR being merged:
- We add an additional `--all-services` (or similar) flag, which makes it a first-class feature and allows people to intentionally invoke the behaviour. (my preferred option)
- We update the flag description for `--service-id` to document that if left blank will return all services.